pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,327 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+)
+
+type eventType uint8
+
+const (
+	groupJoin eventType = iota
+	groupLeave
+
+	// Use IGMPv3 for IGMP query message because it is the main version for IGMP.
+	// TODO: support sending IGMP query with both v2 and v3 messages
+	queryVersion      = 3
+	podInterfaceIndex = "podInterface"
+)
+
+var (
+	workerCount uint8 = 2
+)
+
+type mcastGroupEvent struct {
+	group net.IP
+	eType eventType
+	time  time.Time
+	iface *interfacestore.InterfaceConfig
+}
+
+type GroupMemberStatus struct {
+	group net.IP
+	// localMembers is a set for the local Pod's interface name which has joined in the multicast group.
+	localMembers sets.String
+	lastProbe    time.Time
+	mutex        sync.RWMutex
+}
+
+type worker struct {
+	igmpSnooper  *IGMPSnooper
+	ofClient     openflow.Client
+	groupEventCh chan *mcastGroupEvent
+	staleGroupCh chan net.IP
+	groupCache   cache.Indexer
+	ctx          context.Context
+}
+
+func (w *worker) processGroupEvent(stopCh <-chan struct{}) {
```

Comment:
I found the typical controller pattern is suitable for this case and can simplify the processing of workers:
1. The eventHandler of `mcastGroupEvent` could update `groupCache` synchronously and trigger the reconciliation of corresponding group by adding the group's key (IP) to a workqueue.
2. The consumers of the workqueue get a group, check whether the group still has members (desired state) and decide whether to install flow for the group or remove it based on whether it has installed it (actual state).
3. A separate goroutine enqueues groups that have not been updated for a while periodically.

In this way, `processGroupEvent`, `addGroupMemberStatus`, and `updateGroupMemberStatus` may be merged as a single `syncGroup` method to avoid some redundant code.

Reply:
Not the same. Here the mcastGroupEvent is received from the Pod's PacketIn message, and we don't have a common storage to cache the final state of mcastGroupEvent except for the groupCache inside Antrea. If we use a eventHandler to modify the mcastGroupEvent (I think the eventHandler you mean the PacketIn handler), there should be multiple goroutines accessing the groupCache.

Reply:
I don't get the problem, why can't groupCache contain the final state? why can't groupCache be accessed by multiple goroutines?
```
func eventHandler(event) {
  lock()
  defer unlock()
  addOrUpdateGroupCache(event)
  enqueue(event.Group)
}

func syncGroup(key) {
  group, exists := getGroup(key)
  if !exists {
    return
  }
  if groupHasNoMember(group) {
    if installedGroups.Has(group) {
      deleteGroupFlow(group)
      installedGroups.Delete(group)
    }
    lock()
    unlock()
    deleteGroupCacheIfNoMember(group)
    return
  }
  if !installedGroups.Has(group) {
    installGroupFlow(group)
    installedGroups.Insert(group)
  }
}
```
Can't the above code work?

Reply:
So your point is use the existing goroutine for function `processGroupEvent` to update groupCache according to the groupEvent, and use a separate goroutine to process OpenFlow  entries (name it as goroutine2)? And the goroutines for `clearStaleGroup` and `checkLastMember` are comunicated with the goroutine2. 

The current version is trying simplize the parrallel model to use only one goroutine to write the groupCache, if we change to multiple goroutines to write groupCache, we still need to think of locking status. For example, when syncGroup desides to remove group from groupCache, another gorouine is actually updating its members.

Reply:
Update